src/rust/engine/fs/store/src/snapshot_ops.rs

```diff
@@ -0,0 +1,731 @@
+      parent_path.display(),
+      duplicate_details.join("\n\n")
+    ))
+  }
+
+  async fn snapshot_glob_match(
+    &self,
+    digest: Digest,
+    path_globs: PreparedPathGlobs,
+  ) -> Result<Digest, SnapshotOpsError> {
+    // Split the globs into PathGlobs that can be incrementally matched against individual directory
+    // components.
+    let initial_match_context = UnexpandedSubdirectoryContext {
+      digest,
+      multiple_globs: path_globs.as_expandable_globs().into(),
+    };
+    let mut unexpanded_stack: IndexMap<PathBuf, UnexpandedSubdirectoryContext> =
+      [(PathBuf::new(), initial_match_context)]
+        .iter()
+        .map(|(a, b)| (a.clone(), b.clone()))
+        .collect();
+    let mut partially_expanded_stack: IndexMap<PathBuf, PartiallyExpandedDirectoryContext> =
+      IndexMap::new();
+
+    // 1. Determine all the digests we need to recurse through and modify in order to respect a glob
+    // which may match over multiple contiguous directory components.
+    while let Some((
+      prefix,
+      UnexpandedSubdirectoryContext {
+        digest,
+        multiple_globs,
+      },
+    )) = unexpanded_stack.pop()
+    {
+      // 1a. Extract a single level of directory structure from the digest.
+      let (cur_dir, _metadata) = self
+        .store
+        .load_directory(digest)
+        .await?
+        .ok_or_else(|| format!("directory digest {:?} was not found!", digest))?;
+
+      // NB: When iterating over files, we can remove them from `cur_dir_files` after they are
+      // successfully matched once, hence the `mut` declaration. This is a small
+      // optimization. However, when iterating over directories, different DirWildcard instances
+      // within a single Vec<PathGlob> can result in having multiple different DirWildcard instances
+      // created after matching against a single directory node! So we do *not* mark
+      // `cur_dir_directories` as `mut`.
+      let mut cur_dir_files: IndexMap<PathBuf, remexec::FileNode> = cur_dir
+        .get_files()
+        .to_vec()
+        .into_iter()
+        .map(|file_node| (PathBuf::from(file_node.get_name()), file_node))
+        .collect();
+      let cur_dir_directories: IndexMap<PathBuf, remexec::DirectoryNode> = cur_dir
+        .get_directories()
+        .to_vec()
+        .into_iter()
+        .map(|directory_node| (PathBuf::from(directory_node.get_name()), directory_node))
+        .collect();
+
+      // 1b. Filter files and directories by globs.
+      let MultipleGlobs { include, exclude } = multiple_globs;
+      let mut globbed_files: IndexMap<PathBuf, remexec::FileNode> = IndexMap::new();
+      let mut globbed_directories: IndexMap<PathBuf, remexec::DirectoryNode> = IndexMap::new();
+      let mut todo_directories: IndexMap<PathBuf, Vec<RestrictedPathGlob>> = IndexMap::new();
+
+      // Populate globbed_{files,directories} by iterating over all the globs.
+      for path_glob in include.into_iter() {
```

Comment:
The body of one of these two nested loops (`for path_glob in include.into_iter()` or `for directory_path in matching_directories.into_iter()`) should probably be broken out to take mutable references to the output collections... they account for most of the length of the containing method, and are pretty nested.

Reply:
I've extracted the `for path_glob in include.into_iter()` section into its own struct!